### PR TITLE
SSSP unit tests

### PR DESCRIPTION
**Pull request by YakhtinLeonid**

### CODE REVIEW

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <array>
+
+#include "oneapi/dal/algo/shortest_paths/traverse.hpp"
+#include "oneapi/dal/graph/detail/directed_adjacency_vector_graph_builder.hpp"
+
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:21]: Is this empty blank required by the general testing design? If no, remove.

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            int,
+            std::allocator<char>>(graph_data.get_vertex_count(),
+                                  graph_data.get_edge_count(),
+                                  graph_data.rows.data(),
+                                  graph_data.cols.data(),
+                                  graph_data.edge_weights.data());
+
+        const auto& graph = graph_builder.get_graph();
+
+        auto result_type = nothing_to_calculate
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:119]: to compute

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                               ? optional_results::distances & optional_results::predecessors
+                               : optional_results::distances | optional_results::predecessors;
+        std::allocator<char> alloc;
+
+        const auto shortest_paths_desc =
+            descriptor<float, method::delta_stepping, task::one_to_all, std::allocator<char>>(
+                source,
+                delta,
+                result_type,
+                alloc);
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:129]: Why do we need this?  std alloc is the default type and default constructor as a default value.

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 4]
```diff
@@ -17,9 +17,937 @@
...
+
+protected:
+    std::int64_t vertex_count;
+    std::int64_t edge_count;
+    std::int64_t cols_count;
+    std::int64_t rows_count;
+    int source;
+};
+
+class d_thread_buckets_type : public graph_base_data {
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:62]: d_thread... is not a clear naming

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 5]
```diff
@@ -17,9 +17,937 @@
...
+        edge_count = 0;
+        cols_count = 0;
+        rows_count = 11;
+        source = 0;
+    }
+    std::vector<std::int64_t> rows = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+    std::vector<std::int32_t> cols = {};
+    std::vector<double> edge_weights = {};
+    std::vector<float> distances = { 0,
+                                     std::numeric_limits<float>::infinity(),
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:224]: create an alias or constexpr var with clear short name.

Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:224]: Why std vector? Can we use an array?

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 6]
```diff
@@ -17,9 +17,937 @@
...
+        edge_count = 0;
+        cols_count = 0;
+        rows_count = 11;
+        source = 0;
+    }
+    std::vector<std::int64_t> rows = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+    std::vector<std::int32_t> cols = {};
+    std::vector<int32_t> edge_weights = {};
+    std::vector<float> distances = { 0,
+                                     static_cast<float>(std::numeric_limits<int32_t>::max()),
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:248]: Why cast?

Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:248]: Why distances are float type?

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 7]
```diff
@@ -17,9 +17,937 @@
...
+        max_allocation_size = other.max_allocation_size;
+    }
+
+    template <class U>
+    bool operator==(const LimitedAllocator<U>&) const noexcept {
+        return true;
+    }
+
+    template <class U>
+    bool operator!=(const LimitedAllocator<U>&) const noexcept {
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:646]: Why do you need == and !=

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 8]
```diff
@@ -17,9 +17,937 @@
...
+        35, 73, 94,  99, 34, 39, 11, 39, 47, 73, 54, 93,  54, 25, 75,  95, 51, 83,  76, 18, 79,
+        24, 89, 76,  24, 73, 93, 13, 11, 51, 54, 49, 23,  28, 21, 16,  98, 40, 13,  26, 11, 23
+    };
+    std::vector<float> distances = { 46, 47, 80, 27, 53, 0, 30, 55, 24, 58, 53, 43, 46, 50, 40 };
+};
+
+template <class T>
+struct LimitedAllocator {
+    typedef T value_type;
+    typedef T* pointer;
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:625]: unused?

Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:625]: Check std allocator traits requirements for the allocator and cover the only required things

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 9]
```diff
@@ -17,9 +17,937 @@
...
+
+    bool is_limited;
+    size_t max_allocation_size;
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:628]: Initial value?

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 10]
```diff
@@ -17,9 +17,937 @@
...
+        return static_cast<T*>(pv);
+    }
+
+    void deallocate(T* const p, size_t n) const noexcept {
+        free(p);
+    }
+};
+
+template <typename Type>
+float get_inf_value() {
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:670]: constexpr

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 11]
```diff
@@ -17,9 +17,937 @@
...
+    return std::numeric_limits<Type>::max();
+}
+
+template <>
+float get_inf_value<double>() {
+    return std::numeric_limits<double>::infinity();
+}
+
+class shortest_paths_test {
+    const float kEps = 5e-5;
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:680]: Why do we use this value?

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 12]
```diff
@@ -17,9 +17,937 @@
...
+                if (std::abs(inf_value - distances[index]) > kEps) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    std::vector<float> get_data_from_table(const oneapi::dal::table& table) {
+        auto arr = oneapi::dal::row_accessor<const float>(table).pull();
```
Comment (orrrrtem) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:744]: Why always float?

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    int get_source() const {
+        return source;
+    }
+
+protected:
+    std::int64_t vertex_count;
+    std::int64_t edge_count;
+    std::int64_t cols_count;
+    std::int64_t rows_count;
+    int source;
```
Comment (bysheva) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:56]: Probably source should be int64 as defined in algorithm

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+        const auto result_shortest_paths = dal::preview::traverse(shortest_paths_desc, graph);
+    }
+};
+
+#define SHORTEST_PATHS_BADARG_TEST(name) \
+    TEST_M(shortest_paths_badargs_test, name, "[shortest_paths][badarg]")
+
+SHORTEST_PATHS_BADARG_TEST("Check delta is > 0") {
+    // REQUIRE_THROWS_AS((this->check_shortest_paths<example_graph_type>(0, 0)), invalid_argument);
```
Comment (bysheva) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:133]: Is it obsolete comment? should we remove it?

File: cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+SHORTEST_PATHS_BADARG_TEST("Check source is in graph ") {
+    REQUIRE_THROWS_AS((this->check_shortest_paths<example_graph_type>(5, -2)), invalid_argument);
+    REQUIRE_THROWS_AS((this->check_shortest_paths<example_graph_type>(5, 100)), invalid_argument);
+}
+
+SHORTEST_PATHS_BADARG_TEST("Check empty graph") {
+    REQUIRE_THROWS_AS((this->check_shortest_paths<empty_graph_type>(5, 0)), invalid_argument);
+}
+
+// SHORTEST_PATHS_BADARG_TEST("Check edges are non-negative") {
```
Comment (bysheva) [inline cpp/oneapi/dal/algo/shortest_paths/test/badargs.cpp:149]: Is it obsolete comment? should we remove it?

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_graphs.hpp [hunk 16]
```diff
@@ -0,0 +1,615 @@
...
+    int get_source() const {
+        return source;
+    }
+
+protected:
+    std::int64_t vertex_count;
+    std::int64_t edge_count;
+    std::int64_t cols_count;
+    std::int64_t rows_count;
+    int source;
```
Comment (bysheva) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_graphs.hpp:53]: The same, source is int64

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 17]
```diff
@@ -17,9 +17,337 @@
...
+
+SHORTEST_PATHS_TEST("Custom allocator case") {
+    this->shortest_paths_custom_allocator_check<d_graph_3_graph_type, double>(
+        0.5,
+        true,
+        true,
+        LimitedAllocator<char>());
+}
+
+// SHORTEST_PATHS_TEST("Custom allocator with limited allocations (cannot allocate enough memory)"){
```
Comment (bysheva) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:311]: Obsolete comment?

**Review by bysheva — CHANGES_REQUESTED** [review]

**Review by bysheva — APPROVED** [review]

**Pull request merged**